From: Markus Koschany <apo@debian.org>
Date: Thu, 1 Feb 2024 17:39:38 +0100
Subject: dix-Fix-use-after-free-in-input-device-shutdown

---
 dix/devices.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dix/devices.c b/dix/devices.c
index 8745602..401fe54 100644
--- a/dix/devices.c
+++ b/dix/devices.c
@@ -531,6 +531,7 @@ DisableDevice(DeviceIntPtr dev, BOOL sendevent)
     }
 
     RecalculateMasterButtons(dev);
+    dev->master = NULL;
 
     return TRUE;
 }
