From: Markus Koschany <apo@debian.org>
Date: Thu, 1 Feb 2024 14:26:23 +0100
Subject: Xi-when-creating-a-new-ButtonClass-set-the-number-of

---
 Xi/exevents.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/Xi/exevents.c b/Xi/exevents.c
index 95f776e..e168493 100644
--- a/Xi/exevents.c
+++ b/Xi/exevents.c
@@ -561,6 +561,7 @@ DeepCopyPointerClasses(DeviceIntPtr from, DeviceIntPtr to)
                 to->button = calloc(1, sizeof(ButtonClassRec));
                 if (!to->button)
                     FatalError("[Xi] no memory for class shift.\n");
+                to->button->numButtons = from->button->numButtons;
             }
             else
                 classes->button = NULL;
